pkg/autodiscovery/common/utils/prometheus_kubelet.go

```diff
@@ -28,18 +29,43 @@ func ConfigsForPod(pc *types.PrometheusCheck, pod *kubelet.Pod) []integration.Co
 
 	instances, found := buildInstances(pc, pod.Metadata.Annotations, namespacedName)
 	if found {
-		for _, container := range pod.Status.GetAllContainers() {
-			if !pc.AD.MatchContainer(container.Name) {
-				log.Debugf("Container '%s' doesn't match the AD configuration 'kubernetes_container_names', ignoring it", container.Name)
+		// If `prometheus.io/port` annotation has been provided, let’s keep only the container that declares this port.
+		var portFromAnnotation int
+		var containerPorts map[string]map[int]struct{}
+		if portFromAnnotationString, portFromAnnotationFound := pod.Metadata.Annotations[types.PrometheusPortAnnotation]; portFromAnnotationFound {
+			var err error
+			if portFromAnnotation, err = strconv.Atoi(portFromAnnotationString); err != nil {
+				portFromAnnotationFound = false
+				portFromAnnotation = 0
+			} else {
+				containerPorts = make(map[string]map[int]struct{})
+				for _, containerSpec := range pod.Spec.Containers {
+					containerPorts[containerSpec.Name] = make(map[int]struct{})
+					for _, port := range containerSpec.Ports {
+						containerPorts[containerSpec.Name][port.ContainerPort] = struct{}{}
```

Comment:
Would it make sense to check if `portFromAnnotation` matches container ports here? And then we only need to remember the container name, instead of the full container-port map.

Then, unless I missed something, the check in the main loop can become `if portContainerName != "" && portContainerName != containerStatus.Name { continue }`

Reply:
Hello @vickenty,
I’ve reworked this PR.
Would you mind having a new look on it ?